File: app/src/main/java/com/lam/pedro/util/services/GeofenceBroadcastReceiver.java
package com.lam.pedro.util.services;

import android.app.NotificationManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import com.google.android.gms.location.Geofence;
import com.google.android.gms.location.GeofencingEvent;
import com.lam.pedro.R;

import java.util.List;

public class GeofenceBroadcastReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent != null) {
            GeofencingEvent geofencingEvent = GeofencingEvent.fromIntent(intent);
            if (geofencingEvent.hasError()) {
                // Gestisci l'errore
                int errorCode = geofencingEvent.getErrorCode();
                Log.e("GeofenceReceiver", "Error in geofence: " + errorCode);
                return;
            }

            // Identifica il tipo di transizione (entrata, uscita, permanenza)
            int geofenceTransition = geofencingEvent.getGeofenceTransition();

            if (geofenceTransition == Geofence.GEOFENCE_TRANSITION_ENTER ||
                    geofenceTransition == Geofence.GEOFENCE_TRANSITION_EXIT ||
                    geofenceTransition == Geofence.GEOFENCE_TRANSITION_DWELL) {

                // Ottieni gli ID dei geofence attivati
                List<Geofence> triggeringGeofences = geofencingEvent.getTriggeringGeofences();
                for (Geofence geofence : triggeringGeofences) {
                    String geofenceId = geofence.getRequestId();
                    Log.d("GeofenceReceiver", "Geofence triggered: " + geofenceId);
                }

                // Aggiungi qui la logica per gestire l'evento
                sendNotification(context, geofenceTransition);
            }
        }
    }

    private void sendNotification(Context context, int geofenceTransition) {
        // Esempio: Invio di una notifica
        String transitionType = "";
        switch (geofenceTransition) {
            case Geofence.GEOFENCE_TRANSITION_ENTER:
                transitionType = "Entered geofence";
                break;
            case Geofence.GEOFENCE_TRANSITION_EXIT:
                transitionType = "Exited geofence";
                break;
            case Geofence.GEOFENCE_TRANSITION_DWELL:
                transitionType = "Dwelling in geofence";
                break;
        }

        NotificationManager notificationManager =
                (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, "GeofenceChannel")
                .setSmallIcon(R.drawable.pedro_icon)
                .setContentTitle("Geofence Alert")
                .setContentText(transitionType)
                .setPriority(NotificationCompat.PRIORITY_HIGH);

        if (notificationManager != null) {
            notificationManager.notify(1, builder.build());
        }
    }
}

File: app/src/main/java/com/lam/pedro/util/services/GeofenceBroadcastReceiver.kt
package com.lam.pedro.util.services

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.ui.platform.LocalContext
import com.google.android.gms.location.GeofenceStatusCodes
import com.google.android.gms.location.GeofencingEvent

@RequiresApi(Build.VERSION_CODES.TIRAMISU)
@Composable
fun GeofenceBroadcastReceiver(
    systemAction: String,
    systemEvent: (userActivity: String) -> Unit,
) {
    val TAG = "GeofenceReceiver"
    val context = LocalContext.current
    val currentSystemOnEvent by rememberUpdatedState(systemEvent)

    DisposableEffect(context, systemAction) {
        val intentFilter = IntentFilter(systemAction)
        val broadcast = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                val geofencingEvent = intent?.let { GeofencingEvent.fromIntent(it) } ?: return

                if (geofencingEvent.hasError()) {
                    val errorMessage =
                        GeofenceStatusCodes.getStatusCodeString(geofencingEvent.errorCode)
                    Log.e(TAG, "onReceive: $errorMessage")
                    return
                }
                val alertString = "Geofence Alert :" +
                        " Trigger ${geofencingEvent.triggeringGeofences}" +
                        " Transition ${geofencingEvent.geofenceTransition}"
                Log.d(
                    TAG,
                    alertString
                )
                currentSystemOnEvent(alertString)
            }
        }
        context.registerReceiver(broadcast, intentFilter, Context.RECEIVER_EXPORTED)
        onDispose {
            context.unregisterReceiver(broadcast)
        }
    }
}